TheMET/TheMET/TagListView/TagListView.swift

```diff
@@ -0,0 +1,205 @@
+        self.showMoreTagsButton.frame = CGRect(origin: .zero,
+                                               size: self.showMoreTagsButton.intrinsicContentSize)
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func sizeThatFits(_ size: CGSize) -> CGSize {
+        let estimatedHeight: CGFloat = self.calculateLayout(for: size.width).height
+        return CGSize(width: size.width, height: estimatedHeight)
+    }
+    
+    @objc
+    private func showMoreButtonDidTap() {
+        self.maxNumberOfLines = .max
+    }
+    
+    typealias frameData = (tagViewFrames: [CGRect], showMoreButtonFrame: CGRect?, height: CGFloat)
+    
+    
+    private func calculateLayout(for width: CGFloat) -> frameData {
+        var result: frameData = frameData(tagViewFrames: [], showMoreButtonFrame: .zero, height: 0)
+        guard width > 0 else {
+            return result
+        }
+        guard self.tagViews.count != 0 else {
+            return result
+        }
+        var x: CGFloat = self.offsetByX
+        var y: CGFloat = self.offsetByY
+        let xSpace: CGFloat = self.tagsXSpacing
+        let ySpace: CGFloat = self.tagsYSpacing
+        var lineNumber: Int = 0
+    throughTags: for (tagViewIndex, tagView) in tagViews.enumerated() {
+        if lineNumber < self.maxNumberOfLines - 1 {
+            if tagView.bounds.width > width - x {
+                lineNumber += 1
+                y += tagView.bounds.height + ySpace
+                x = self.offsetByX
+            }
+            if lineNumber == self.maxNumberOfLines - 1,
+               tagView.bounds.width > self.showMoreTagsButton.bounds.width {
+                
+            } else {
+                result.tagViewFrames.append(CGRect(x: x, y: y, width: tagView.bounds.width, height: tagView.bounds.height))
+                x += tagView.bounds.width + xSpace
+            }
+        } else {
+            let restTagsWidth = self.tagsWidth(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace)
+            if restTagsWidth < width {
+                let restTagsFrames:[CGRect] = self.setRestTagsFrames(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace, ySpace: ySpace, x: x, y: y)
+                result.tagViewFrames.append(contentsOf: restTagsFrames)
+                break throughTags
+            } else {
+                let restTagsAndShowMoreButtonFrames = self.setRestTagsAndShowMoreButtonFrames(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace, ySpace: ySpace, x: x, y: y)
+                result.tagViewFrames.append(contentsOf: restTagsAndShowMoreButtonFrames.tagViewFrames)
+                result.showMoreButtonFrame = restTagsAndShowMoreButtonFrames.showMoreButton
+                break throughTags
+            }
+        }
+    }
+        self.totalHeight = self.tagViews[0].bounds.height * CGFloat(lineNumber + 1) + ySpace * CGFloat(lineNumber)
```

Comment:
`totalHeight` must not be set in this method. This method must only calculate a `height` for the provided width. `totalHeight` supposed to be set in `layoutTagViews` method.

Reply:
fixed